README.md
# hashmap-client
a javascript client library for interacting with [hashmap](https://github.com/nomasters/hashmap) server and payloads.

[![CircleCI][1]][2] [![Known Vulnerabilities][3]][4] [![Coverage Status][5]][6] [![npm version][7]][8]

[1]: https://circleci.com/gh/nomasters/hashmap-client-js.svg?style=svg
[2]: https://circleci.com/gh/nomasters/hashmap-client-js
[3]: https://snyk.io/test/github/nomasters/hashmap-client-js/badge.svg
[4]: https://snyk.io/test/github/nomasters/hashmap-client-js
[5]: https://coveralls.io/repos/github/nomasters/hashmap-client-js/badge.svg?branch=master
[6]: https://coveralls.io/github/nomasters/hashmap-client-js?branch=master
[7]: https://img.shields.io/npm/v/hashmap-client.svg
[8]: https://www.npmjs.com/package/hashmap-client

## Summary

This is a simple client for generating, interacting with, submitting, and fetching hashmap payloads. If you are new to `hashmap`, you can read more about [it in its github page](https://github.com/nomasters/hashmap).

NOTE: `hashmap` is in early alpha. It has not been properly reviewed for security and privacy leaks and should only be used for experimental purposes. Large and small breaking changes may be introduces while in early alpha.

Install Instructions

```
npm install hashmap-client
```

The purpose of this library is to make interacting with hashmap payloads easy and predictable. Under the hood, this library runs through a series of validity checks on the signature, endpoint to public key verification, and formatting requirements.

### Initializing a payload

To interact with a payload, initiate a new instance of a payload.

```
let payload = new hashmap.Payload();
```

You can set a global default `ServerURI` with the `setServerURI(uri)` function

```
hashmap.setServerURI('https://prototype.hashmap.sh')
```

You can optionally initialize a payload with the `uri` and `endpoint` for a hashmap key-value store on initialization of a new instance of payload. Passing in a uri for a payload overrides the global default for this specific instance.

```
let opts = {
    uri: "https://prototype.hashmap.sh",
    endpoint: "2DrjgbD6zUx2svjd4NcXfsTwykspqEQmcC2WC7xeBUyPcBofuo",
}
let payload = new hashmap.Payload(opts);
```

### Getting a payload from a hashmap server

To get a hashmap payload from a specific endpoint, either initialize a payload with the `uri` and `endpoint` as outlined above, or pass them in as arguments in the `.get()` method.

The `get()` method returns a promise that resolves to a payload object or an error.


```
let endpoint = '2DrjgbD6zUx2svjd4NcXfsTwykspqEQmcC2WC7xeBUyPcBofuo';
let payload  = new hashmap.Payload();

payload.get(endpoint)
	.then(resp => console.log(resp))
	.catch(err => console.log(err))
```

### Generating and Posting a payload to a hashmap server

To post a payload, you will need:

- the `uri` for the hashmap server
- a base64 encoded `ed25519` private key (for signing the payload)
- a string formatted message to include in the payload (no larger than 512 bytes)

The generator takes care of the rest. 

`hashmap-client` also includes an `ed25519` private key generator that returns the properly formatted base64 string


```
// Generate a Private Key (don't share this)
let key = hashmap.genNaClSignPrivKey();
let message = 'hello, world';

// add uri to options
let opts = { uri: "https://prototype.hashmap.sh" }

// initialize Payload with options
let payload = new hashmap.Payload(opts)

// generate the signed hashmap payload
payload.generate(key, message)

payload.post()
.then(resp => console.log(resp))
.catch(err => console.log(err))

```

NOTE: Post the payload as soon as possible after generating it. Hashmap server checks the signed timestamp included with the payload and rejects payloads outside of a timestamp drift threshold.



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.1.1] - 2018-08-31
### Fixed
- fixed issue implimentation of `getBlake2b256MultiHash()` to properly hash and encode publicKey to to blake2b256 in multihash format. This was causing a false mismatch on the get request validating the URL to the payload.

## [0.1.0] - 2018-08-31
### Added
- added `setServerURI` and `getServerURI` functions to set global hashmap server defaults


## [0.0.2] - 2018-08-30
### Added
- integrated with ci and coverage tools
- added status badges
- added (this) CHANGELOG


## [0.0.1] - 2018-08-30
### Added
- initial commits of hashmap

test.ts
import * as hashmap from './dist/index';
import { expect } from 'chai';
import * as nacl from 'tweetnacl';
import * as multihash from 'multihashes';
import 'mocha';
import * as nock from 'nock';

describe('Constants', () => {
    const testTable = [
        ['maxMessageBytes', hashmap.maxMessageBytes, 512],
        ['defaultSigMethod', hashmap.defaultSigMethod, 'nacl-sign-ed25519'],
        ['dataTTLDefault', hashmap.dataTTLDefault, 86400],
        ['dataTTLMax', hashmap.dataTTLMax, 604800],
        ['version', hashmap.version, '0.0.1'],
    ]
    for (let i=0; i<testTable.length; i++) {
        let t = testTable[i]
        it(`should match ${t[0]} to ${t[2]}`, () => {
            expect(t[1]).to.equal(t[2]);
        });
    }
});

describe('Private Key Generator', () => {
    const key  = hashmap.genNaClSignPrivKey();
    it('should be be the proper length', () => {
        expect(key.length).to.equal(88);
    });
    it('should be a string', () => {
        expect(key).to.be.a('string');
    });
    it('should be able to sign and open a message', () => {
        const privKey = Buffer.from(key, 'base64');
        const pubKey  = privKey.slice(32, 64);
        const message = Buffer.from('test message', 'ascii')
        const signedMessage = nacl.sign(message, privKey)
        const openedMessage = Buffer.from(nacl.sign.open(signedMessage, pubKey))
        expect(openedMessage.toString('base64')).to.equal(message.toString('base64'));
    });
});

describe('Public Key', () => {
    const key  = hashmap.genNaClSignPrivKey();
    it('should valid base64 string from private key', () => {
        const privKey = Buffer.from(key, 'base64');
        const pubKey  = privKey.slice(32, 64).toString('base64');
        const pk      = hashmap.getEd25519PubkeyFromPrivateKey(key);
        expect(pk).to.be.equal(pubKey);
    });
    it('should be able to encode and decode from multihash', () => {
        const pk        = hashmap.getEd25519PubkeyFromPrivateKey(key);
        const pkHash    = hashmap.getBlake2b256MultiHash(pk)
        const multih    = multihash.fromB58String(pkHash)
        const mh        = multihash.decode(multih)
        expect(mh.digest.toString('base64')).to.equal(pk)
    }); 
});

describe('Unix Nano Time Stamp Generator', () => {
    it('should be a number', () => {
        const timestamp  = hashmap.unixNanoNow();
        expect(timestamp).to.be.a('number');
    });
    it('should be accurate to at least Date.now() (1/1000 of a second)', () => {
        const timestamp  = Math.floor(hashmap.unixNanoNow() / 1000000);
        const unixTime = Math.floor(Date.now());
        expect(timestamp).to.equal(unixTime);
    });
});

describe('Payload', () => {
    const defaultPrivKey = hashmap.genNaClSignPrivKey();
    const defaultPubkey  = hashmap.getEd25519PubkeyFromPrivateKey(defaultPrivKey);
    const uri            = 'https://prototype.hashmap.sh'
    const endpoint       = '2DrjgbD6zUx2svjd4NcXfsTwykspqEQmcC2WC7xeBUyPcBofuo'
    const defaultMessage = 'test'
    
    it('should initialize without uri options', () => {
        let p = new hashmap.Payload()
        expect(p.uri).to.be.undefined
    });
    it('should initialize without endpoint options', () => {
        let p = new hashmap.Payload()
        expect(p.endpoint).to.be.undefined
    });
    it('should initialize with uri options', () => {
        let opts = { uri: uri, endpoint: endpoint }
        let p = new hashmap.Payload(opts)
        expect(p.uri).to.equal(opts.uri)
    });
    it('should initialize with endpoint options', () => {
        let opts = { uri: uri, endpoint: endpoint }
        let p = new hashmap.Payload(opts)
        expect(p.endpoint).to.equal(opts.endpoint)
    });
    it('should generate a signed payload', () => {
        let p = new hashmap.Payload()
        p.generate(defaultPrivKey, defaultMessage)
        expect(p.raw.pubkey).to.be.a('string')
    });
    it('should generate a signed payload with a custom ttl', () => {
        let p = new hashmap.Payload()
        p.generate(defaultPrivKey, defaultMessage, { ttl: 5 })
        expect(p.raw.pubkey).to.be.a('string')
    });
    it('should reject a signed payload attempt with ttl greater than max', () => {
        let p = new hashmap.Payload()
        let opts = { ttl: hashmap.dataTTLMax + 1 }
        expect(() => p.generate(defaultPrivKey, defaultMessage, opts)).to.throw()
    });
    it('should reject a signed payload attempt when the message is too large', () => {
        let p = new hashmap.Payload()
        let message = `fdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
        deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
        deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
        deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
        deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
        deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
        deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
        deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef`
        expect(() => p.generate(defaultPrivKey, message)).to.throw()
    });
    it('should import a properly formatted json payload', () => {
        let p = new hashmap.Payload()
        let jsonPayload = p.generate(defaultPrivKey, defaultMessage, 'test')
        let p2 = new hashmap.Payload()
        p2.import(jsonPayload)
        expect(p2.getMessage()).to.equal(defaultMessage)
    });
    it('should reject an improperly formatted json payload', () => {
        let jsonPayload = `{"total":"fail"}`
        let p = new hashmap.Payload()
        expect(() => p.import(jsonPayload)).to.throw()
    });
    it('should get a payload from a hashmap server', () => {
        let g    = new hashmap.Payload()
        let body = g.generate(defaultPrivKey, defaultMessage)

        nock(uri)
        .get('/' + hashmap.getBlake2b256MultiHash(defaultPubkey))
        .reply(200, body);

        let opts = {
            uri: uri,
            endpoint: hashmap.getBlake2b256MultiHash(defaultPubkey),
        }
        let p = new hashmap.Payload(opts)
        return p.get().then(r => expect(g.raw.pubkey).to.equal(r.pubkey))
    });
    it('should post a payload to a hashmap server', () => {
        nock(uri)
        .post('/')
        .reply(200, { endpoint: hashmap.getBlake2b256MultiHash(defaultPubkey) });

        let opts = { uri: uri }
        let p = new hashmap.Payload(opts)
        p.generate(defaultPrivKey, defaultMessage)
        return p.post()
            .then(r => {
                expect(r.endpoint).to.equal(hashmap.getBlake2b256MultiHash(defaultPubkey))
            })
    });


});

index-test.ts
import * as hashmap from './dist/index';
import * as nacl from 'tweetnacl';
import * as multihash from 'multihashes';
import * as blake from 'blakejs';
import * as nock from 'nock';
import * as chai from 'chai';
import * as chaiAsPromised from 'chai-as-promised';
import 'mocha';

chai.use(chaiAsPromised);
const expect = chai.expect;

describe('Constants', () => {
    const testTable = [
        ['maxMessageBytes', hashmap.maxMessageBytes, 512],
        ['defaultSigMethod', hashmap.defaultSigMethod, 'nacl-sign-ed25519'],
        ['dataTTLDefault', hashmap.dataTTLDefault, 86400],
        ['dataTTLMax', hashmap.dataTTLMax, 604800],
        ['version', hashmap.version, '0.0.1'],
    ]
    for (let i=0; i<testTable.length; i++) {
        let t = testTable[i]
        it(`should match ${t[0]} to ${t[2]}`, () => {
            expect(t[1]).to.equal(t[2]);
        });
    }
});

describe('Private Key Generator', () => {
    const key  = hashmap.genNaClSignPrivKey();
    it('should be be the proper length', () => {
        expect(key.length).to.equal(88);
    });
    it('should be a string', () => {
        expect(key).to.be.a('string');
    });
    it('should be able to sign and open a message', () => {
        const privKey = Buffer.from(key, 'base64');
        const pubKey  = privKey.slice(32, 64);
        const message = Buffer.from('test message', 'ascii')
        const signedMessage = nacl.sign(message, privKey)
        const openedMessage = Buffer.from(nacl.sign.open(signedMessage, pubKey))
        expect(openedMessage.toString('base64')).to.equal(message.toString('base64'));
    });
});

describe('Public Key', () => {
    const key  = hashmap.genNaClSignPrivKey();
    it('should valid base64 string from private key', () => {
        const privKey = Buffer.from(key, 'base64');
        const pubKey  = privKey.slice(32, 64).toString('base64');
        const pk      = hashmap.getEd25519PubkeyFromPrivateKey(key);
        expect(pk).to.be.equal(pubKey);
    });
    it('should be able to encode and decode from multihash', () => {
        const pk        = hashmap.getEd25519PubkeyFromPrivateKey(key);
        const pkHash    = hashmap.getBlake2b256MultiHash(pk)
        const multih    = multihash.fromB58String(pkHash)
        const mh        = multihash.decode(multih)
        const hash      = Buffer.from(blake.blake2b(Buffer.from(pk, 'base64'), null, 32)).toString('base64')

        expect(mh.digest.toString('base64')).to.equal(hash)
    }); 
});

describe('Unix Nano Time Stamp Generator', () => {
    it('should be a number', () => {
        const timestamp  = hashmap.unixNanoNow();
        expect(timestamp).to.be.a('number');
    });
    it('should be accurate to at least Date.now() (1/1000 of a second)', () => {
        const timestamp  = Math.floor(hashmap.unixNanoNow() / 10000000);
        const unixTime = Math.floor(Date.now() / 10);
        expect(timestamp).to.equal(unixTime);
    });
});

describe('ServerURI', () => {
    it('should work with setter and getter functions', () => {
        const uri = 'https://prototype.hashmap.sh'
        hashmap.setServerURI(uri)
        expect(hashmap.getSeverURI()).to.equal(uri)
        hashmap.setServerURI('')
    });
})

describe('Payload', () => {
    const defaultPrivKey = hashmap.genNaClSignPrivKey();
    const defaultPubkey  = hashmap.getEd25519PubkeyFromPrivateKey(defaultPrivKey);
    const uri            = 'https://prototype.hashmap.sh'
    const endpoint       = hashmap.getBlake2b256MultiHash(defaultPubkey)
    const defaultMessage = 'test'
    
    it('should initialize without uri options', () => {
        let p = new hashmap.Payload()
        expect(p.uri).to.be.undefined
    });
    it('should initialize with serverURI', () => {
        hashmap.setServerURI(uri)
        let p = new hashmap.Payload()
        hashmap.setServerURI('')
        expect(p.uri).to.equal(uri)
    });
    it('should initialize without endpoint options', () => {
        let p = new hashmap.Payload()
        expect(p.endpoint).to.be.undefined
    });
    it('should initialize with uri options', () => {
        let opts = { uri: uri, endpoint: endpoint }
        let p = new hashmap.Payload(opts)
        expect(p.uri).to.equal(opts.uri)
    });
    it('should initialize with endpoint options', () => {
        let opts = { uri: uri, endpoint: endpoint }
        let p = new hashmap.Payload(opts)
        expect(p.endpoint).to.equal(opts.endpoint)
    });
    it('should generate a signed payload', () => {
        let p = new hashmap.Payload()
        p.generate(defaultPrivKey, defaultMessage)
        expect(p.raw.pubkey).to.be.a('string')
    });
    it('should generate a signed payload with a custom ttl', () => {
        let p = new hashmap.Payload()
        p.generate(defaultPrivKey, defaultMessage, { ttl: 5 })
        expect(p.raw.pubkey).to.be.a('string')
    });
    it('should generate a signed payload with null message', () => {
        let p = new hashmap.Payload()
        p.generate(defaultPrivKey)
        expect(p.raw.pubkey).to.be.a('string')
    });
    it('should reject a signed payload attempt with ttl greater than max', () => {
        let p = new hashmap.Payload()
        let opts = { ttl: hashmap.dataTTLMax + 1 }
        expect(() => p.generate(defaultPrivKey, defaultMessage, opts)).to.throw()
    });
    it('should reject a signed payload attempt when the message is too large', () => {
        let p = new hashmap.Payload()
        let message = `fdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
        deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
        deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
        deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
        deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
        deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
        deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef
        deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef`
        expect(() => p.generate(defaultPrivKey, message)).to.throw()
    });
    it('should import a properly formatted json payload', () => {
        let p = new hashmap.Payload()
        let jsonPayload = p.generate(defaultPrivKey, defaultMessage, 'test')
        let p2 = new hashmap.Payload()
        p2.import(jsonPayload)
        expect(p2.getMessage()).to.equal(defaultMessage)
    });
    it('should reject an improperly formatted json payload', () => {
        let jsonPayload = `{"total":"fail"}`
        let p = new hashmap.Payload()
        expect(() => p.import(jsonPayload)).to.throw()
    });
    it('should get a payload from a hashmap server', () => {
        let g    = new hashmap.Payload()
        let body = g.generate(defaultPrivKey, defaultMessage)

        nock(uri)
        .get('/' + hashmap.getBlake2b256MultiHash(defaultPubkey))
        .reply(200, body);

        let opts = {
            uri: uri,
            endpoint: hashmap.getBlake2b256MultiHash(defaultPubkey),
        }
        let p = new hashmap.Payload(opts)
        return expect(p.get()).to.eventually.be.fulfilled;
    });
    it('should get a payload to a hashmap serve with endpoint and uri passed to the method', () => {
        let g    = new hashmap.Payload()
        let body = g.generate(defaultPrivKey, defaultMessage)

        nock(uri)
        .get('/' + hashmap.getBlake2b256MultiHash(defaultPubkey))
        .reply(200, body);

        let p = new hashmap.Payload()
        return expect(p.get(endpoint, uri)).to.eventually.be.fulfilled;
    });

    it('should reject a get request without an endpoint', () => {
        let opts = { uri: uri }
        let p = new hashmap.Payload(opts)
        return expect(p.get()).to.eventually.be.rejected;
    });
    it('should reject a get request without a uri', () => {
        let opts = { endpoint: endpoint }
        let p = new hashmap.Payload(opts)
        return expect(p.get()).to.eventually.be.rejected;
    });
    it('should reject a get request from an endpoint to pubkey mismatch', () => {
        const badKey = hashmap.genNaClSignPrivKey();

        let g    = new hashmap.Payload()
        let body = g.generate(badKey, defaultMessage)

        nock(uri)
        .get('/' + hashmap.getBlake2b256MultiHash(defaultPubkey))
        .reply(200, body);

        let opts = {
            uri: uri,
            endpoint: hashmap.getBlake2b256MultiHash(defaultPubkey),
        }
        let p = new hashmap.Payload(opts)
        return expect(p.get()).to.eventually.be.rejected;
    });

    it('should post a payload to a hashmap server', () => {
        nock(uri)
        .post('/')
        .reply(200, { endpoint: hashmap.getBlake2b256MultiHash(defaultPubkey) });

        let opts = { uri: uri }
        let p = new hashmap.Payload(opts)
        p.generate(defaultPrivKey, defaultMessage)
        return expect(p.post()).to.eventually.be.fulfilled;
    });
    it('should post a payload to a hashmap serve with uri passed to post method', () => {
        nock(uri)
        .post('/')
        .reply(200, { endpoint: hashmap.getBlake2b256MultiHash(defaultPubkey) });
        let p = new hashmap.Payload()
        p.generate(defaultPrivKey, defaultMessage)
        return expect(p.post(uri)).to.eventually.be.fulfilled;
    });
    it('should reject a post request without a raw payload', () => {
        let p = new hashmap.Payload()
        return expect(p.post(uri)).to.eventually.be.rejected;
    });
    it('should reject a post request without a uri', () => {
        let p = new hashmap.Payload()
        return expect(p.post()).to.eventually.be.rejected;
    });

    it('should reject an invalid signature', () => {
        let p = new hashmap.Payload()
        p.generate(defaultPrivKey, defaultMessage)
        p.raw.sig = 'fail' + p.raw.sig
        expect(() => p.validateSig()).to.throw()
    });
});

examples/example.js
const hashmap = require('../dist/index.js');

// This code expects a base64 encoded ed25519 private key to 
// be passed in as an arg.
// var args = process.argv.slice(2);
// var key = args[0]

var opts1 = {
    uri: "https://prototype.hashmap.sh",
    endpoint: "2DrjgbEoLCSYbfMBLJhQnatJP9TUKBGRumksAD1g5noGBMD5sk",
}

console.log("ierNf204JhCfuG7cAztAuC2EEw70X9atTGMmdGM9rgE=")


let key = 'xUCiZL1bBx4HfSWohe4m9PaSHcWQ7c7dIEiluQHxw25IPcCO1GNWONn4I+M9DgbWl81ETgGsD+itIMKeTyLXeg=='

// post payload example
var opts2 = { uri: "https://prototype.hashmap.sh" }
p2 = new hashmap.Payload(opts2)
p2.generate(key, "hello, world it is: " + hashmap.unixNanoNow())
p2.post()
.then(resp => console.log(resp))
.catch(err => console.log(err))


// get payload example
p1 = new hashmap.Payload(opts1)
p1.get()
.then(payload => console.log(p1.getMessage()))
.catch(err => console.log(err))


index.ts
import * as nacl from 'tweetnacl';
import * as rp from "request-promise";
import * as BigInt from 'big-integer';
import * as multihash from 'multihashes';
import * as blake from 'blakejs';

export const maxMessageBytes: number  = 512
export const defaultSigMethod: string = 'nacl-sign-ed25519'
export const dataTTLDefault: number   = 86400  // 1 day in seconds
export const dataTTLMax: number       = 604800 // 1 week in seconds
export const version: string          = '0.0.1'

// globally scoped variable for uri
var ServerURI: string = ''

// setServerURI sets the private variable ServerURI
export function setServerURI(uri: string) {
    ServerURI = uri
}

// setServerURI gets the private variable ServerURI
export function getSeverURI() {
    return ServerURI
}

// unixNanoNow is the equivelant of the go time.Now().UnixNano()
// this function was modified from the one in `nano-time` but
// all values are scoped inside the function and this returns
// a native JS Number type instead of a string like in `nano-time`
export function unixNanoNow() {
    let n = process.hrtime();
    let m = new Date().getTime();
    let d = process.hrtime(n);
    return BigInt(m).times(1e6).add(BigInt(d[0]).times(1e9).plus(d[1])).valueOf();
}

// genNaClSignPrivKey generates a new ed25519 private key and returns
// it as a base64 encoded string
export function genNaClSignPrivKey() {
    const keypair = nacl.sign.keyPair()
    return Buffer.from(keypair.secretKey).toString('base64')
}

// getEd25519PubkeyFromPrivateKey takes a base64 encoded string of the 
// private key and returns a base64 encoded string of the public key
export function getEd25519PubkeyFromPrivateKey(privateKey: string) {
    const privKey = Buffer.from(privateKey, 'base64');
    return privKey.slice(32,64).toString('base64');
}

// getBlake2b256MultiHash takes a base64 encoded string of the public key
// and returns a base58 encoded multihash in blake2b256 formatting
export function getBlake2b256MultiHash(publicKey: string) {
    const pubKey = Buffer.from(publicKey, 'base64');
    const hash = Buffer.from(blake.blake2b(pubKey, null, 32))
    return multihash.toB58String(multihash.encode(hash, 'blake2b-256'))
}

// PayloadOptions is the interface used for the Payload Constructor
export interface PayloadOptions {
    uri: string
    endpoint: string
}

// GenerateOptions is the interface used for the generate method
export interface GenerateOptions {
    ttl: number
}

// PayloadObject is the interface description of the JSON Payload
// Parsed into an object
export interface PayloadObject {
    data: string
    sig: string
    pubkey: string
}

// Payload is the primary class used by the hashmap client. It contains
// methods for getting, posting, validating, and analyzing hashmap payloads
export class Payload {
    public uri: string
    public endpoint: string
    public raw: PayloadObject

    // the constructor takes an optional uri and endpoint to be used
    // by get and post methods
    constructor(opts? : PayloadOptions) {
        if (ServerURI !== '') this.uri = ServerURI
        if (opts && opts.uri) this.uri = opts.uri
        if (opts && opts.endpoint) this.endpoint = opts.endpoint
    }

    // generate takes a base64 encoded key, a message string, and opts object
    // and creates a properly formatted and signed payload
    // it returns a JSON encoded string and sets the class
    // internal state for use with other methods
    public generate(key: string, message=' ', opts? : GenerateOptions) {
        let ttl = dataTTLDefault
        if (opts && opts.ttl) {
            ttl = opts.ttl
        }
        if (ttl > dataTTLMax) {
            throw new Error('invalid ttl, exceeds max')
        }
        const data = {
            message: Buffer.from(message, 'ascii').toString('base64'),
            timestamp: unixNanoNow(),
            sigMethod: defaultSigMethod,
            version: version,
            ttl: ttl,
        }
        // dataBytes takes a byte buffer of the data object that has been stringified
        const dataBytes = Buffer.from(JSON.stringify(data), 'ascii')
        const privKey = Buffer.from(key, 'base64');
        const pubKey = privKey.slice(32,64)
        const signedMessage = nacl.sign(dataBytes, privKey)
        const sig = Buffer.from(signedMessage.slice(0,64))

        const p = {
            data: dataBytes.toString('base64'),
            pubkey: pubKey.toString('base64'),
            sig: sig.toString('base64'),
        }
        this.validate(p)
        return JSON.stringify(p)
    }

    // get takes an endpoint and uri string and sends a get request
    // to a hashmap uri and validates the payload. It returns a promise
    // that resolves to a json formatted response
    public get(endpoint?: string, uri?: string) {
        if (!endpoint && !this.endpoint) { 
            return Promise.reject(new Error("missing endpoint")) 
        }
        if (endpoint) this.endpoint = endpoint
        if (!uri && !this.uri) { 
            return Promise.reject(new Error("missing uri")) 
        }
        if (uri) this.uri = uri
        const opts = {
            uri: this.uri + '/' + this.endpoint,
            json: true
        }
        return rp(opts)
            .then(resp => {
                this.validate(resp)
                if (this.endpoint !== getBlake2b256MultiHash(this.raw.pubkey)) {
                    throw new Error('endpoint to pubkey mismatch')
                }
                return resp
            })
    }

    // post takes a uri and posts the raw payload data to a hashmap server.
    // It returns a promise the resolves to the json body of the endpoint updated
    public post(uri?: string) {
        if (!uri && !this.uri) { return Promise.reject(new Error("missing uri")) }
        if (!this.raw) { return Promise.reject(new Error("missing payload")) }
        if (uri) this.uri = uri
        const opts = {
            uri: this.uri,
            method: 'POST',
            body: this.raw,
            json: true,
        }
        return rp(opts)
    }

    // import takes a raw JSON string of a payload, and validates the payload
    public import(rawjson: string) { this.validate(JSON.parse(rawjson)) }

    // validate takes a payloadObject sets and validates a series of checks
    // proper including base64 encoding of values, message requirements, and
    // ensures that the signature is valid based on the message and pubkey
    public validate(p: PayloadObject, opts={}) {
        this.raw     = p
        this.getDataBytes()
        this.getSigBytes()
        this.getPubkeyBytes()
        this.validateMessage()
        this.validateSig()
    }

    // validateMessage checks that the length of bytes doesn't exceed the
    // allowable maximum.
    public validateMessage() {
        if (this.getMessageBytes().length > maxMessageBytes) { 
            throw new Error("message length exceeds max threshold")
        }
    }

    // validateSig concats the sigBytes and the dataBytes and runs the 
    // nacl.sign.open method to ensure the validation of the data is in place
    public validateSig() {
        const sd = Buffer.concat([this.getSigBytes(), this.getDataBytes()])
        if (!nacl.sign.open(sd, this.getPubkeyBytes())) { 
            throw new Error("signature validation failed")
        }
    }

    // getPubkeyBytes returns a bytes buffer for payload pubkey
    public getPubkeyBytes() {
        return Buffer.from(this.raw.pubkey, 'base64');
    }

    // getSigBytes returns a bytes buffer for payload sig
    public getSigBytes() {
        return Buffer.from(this.raw.sig, 'base64');
    }

    // getDataBytes returns a bytes buffer for payload data
    public getDataBytes() {
        return Buffer.from(this.raw.data, 'base64');
    }

    // getData returns the payload data as an object
    public getData() { 
        return JSON.parse(this.getDataBytes().toString('ascii')) 
    }

    // getMessageBytes returns a bytes buffer for Payload.Data.Message
    public getMessageBytes() { 
        return Buffer.from(this.getData().message, 'base64') 
    }

    // getMessage returns the Payload.Data.Message as a string
    public getMessage() { 
        return this.getMessageBytes().toString('ascii') 
    }
}